src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,164 @@
+
+/-- The scalar multiplication depends on the point `x : M`. -/
+def has_scalar (x : M) : has_scalar C^∞⟮I, M; 𝕜⟯ 𝕜 :=
+{ smul := λ f k, f x * k }
+
+lemma scalar_def {x : M} {f : C^∞⟮I, M; 𝕜⟯} {k : 𝕜} :
+  @has_scalar.smul C^∞⟮I, M; 𝕜⟯ 𝕜 (has_scalar I M x) f k = f x * k := rfl
+
+/-- The scalar multiplication defined above gives rise to a module structure. -/
+def module (x : M) : module C^∞⟮I, M; 𝕜⟯ 𝕜 :=
+{ one_smul := λ k, one_mul k,
+  mul_smul := λ f g k, mul_assoc _ _ _,
+  smul_add := λ f g k, mul_add _ _ _,
+  smul_zero := λ f, mul_zero _,
+  add_smul := λ f g k, add_mul _ _ _,
+  zero_smul := λ f, zero_mul _,
+  ..point_derivation.has_scalar I M x }
+
+lemma is_scalar_tower (x : M) :
+  @is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ 𝕜 _ (has_scalar I M x) _ :=
+{ smul_assoc := λ k f h, by { simp only [scalar_def, algebra.id.smul_eq_mul,
+    smooth_map.coe_smul, pi.smul_apply, mul_assoc]} }
+
+end point_derivation
+
+/-- The derivations at a point of a manifold. Some regard this as a possible definition of the
+tangent space -/
+@[reducible] def point_derivation (x : M) :=
+  @derivation 𝕜 C^∞⟮I, M; 𝕜⟯ _ _ _ 𝕜 _ (point_derivation.module I M x) _
+    (point_derivation.is_scalar_tower I M x)
+
+/-- The total bundle of point derivations. -/
+def derivation_bundle := Σ x : M, point_derivation I M x
+
+variables {I M}
+
+/-- The inclusion map of derivations at a point into the total bundle. -/
+def derivation_inclusion {x : M} (v : point_derivation I M x) : derivation_bundle I M :=
+sigma.mk x v
+
+instance [inhabited M] : inhabited (derivation_bundle I M) :=
+⟨derivation_inclusion (0 : point_derivation I M (default M))⟩
+
+section
+
+/- Why do I need to rewrite extensionality rules for reducible defs? -/
+namespace point_derivation
+
+variables {I} {M} {x y : M} {v w : point_derivation I M x} (f g : C^∞⟮I, M; 𝕜⟯) (r : 𝕜)
+
+lemma coe_injective (h : ⇑v = w) : v = w :=
+@derivation.coe_injective 𝕜 _ C^∞⟮I, M; 𝕜⟯ _ _ 𝕜 _ (point_derivation.module I M x) _
+  (point_derivation.is_scalar_tower I M x) v w h
+
+@[ext] theorem ext (h : ∀ f, v f = w f) : v = w :=
+coe_injective $ funext h
+
+theorem hext {u : point_derivation I M y} (h1 : x = y) (h2 : ∀ f, v f = u f) : v == u :=
+by { cases h1, rw heq_iff_eq, ext, exact h2 f }
+
+end point_derivation
+
+end
+
+section
+
+variables {I} {M} (X Y : derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯)
+  (f g : C^∞⟮I, M; 𝕜⟯) (r : 𝕜)
+
+namespace derivation
+
+/-- Evaluation of a global derivation at a point, giving a point derivation in the most natural
+possible way. -/
+def eval_map (x : M) : point_derivation I M x :=
+{ to_fun := λ f, (X f) x,
+  map_add' := λ f g, by { rw derivation.map_add, refl },
+  map_smul' := λ f g, by { rw [derivation.map_smul, algebra.id.smul_eq_mul], refl },
+  leibniz' := λ h k, by { dsimp only [], rw [derivation.leibniz, algebra.id.smul_eq_mul], refl } }
+
+/-- The evaluation is a linear map. -/
+def eval_at (x : M) : (derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^⊤⟮I, M; 𝕜⟯) →ₗ[𝕜] point_derivation I M x :=
```

Comment:
I guess mathematically, in this approach if we fix a point `x : M`, we have two representations of the algebra of smooth functions `C^∞⟮I, M; 𝕜⟯`, namely the algebra itself and `𝕜` and furthermore we have a morphism of `C^∞⟮I, M; 𝕜⟯`-algebras: `C^∞⟮I, M; 𝕜⟯ →ₐ[C^∞⟮I, M; 𝕜⟯] 𝕜`, namely evaluation at `x`.

If that setup can be made to work, then we can drop `eval_map` and `eval_at` in favour of what I call `derivation.comp` below:
```lean
section foo

variables (R A M N : Type*) [comm_semiring R] [comm_semiring A] [algebra R A]
variables [add_cancel_comm_monoid M] [module A M] [module R M] [is_scalar_tower R A M]
variables [add_cancel_comm_monoid N] [module A N] [module R N] [is_scalar_tower R A N]

@[simp] lemma derivation.mk_coe (f : A →ₗ[R] M) (h) :
  ((⟨f, h⟩ : derivation R A M) : A → M) = f :=
rfl

def derivation.comp (f : M →ₗ[A] N) : derivation R A M →ₗ[R] derivation R A N :=
{ to_fun    := λ D,
  { leibniz'  := λ a b, by simp,
    .. (f : M →ₗ[R] N).comp (D : A →ₗ[R] M), },
  map_add'  := λ D₁ D₂, by ext; simp,
  map_smul' := λ r D, by ext; simp, }

end foo
```

Reply:
I think I understand what you mean and what is going on mathematically, but I don't know enough algebra here: is this construction that you point out also useful in other contexts? It is beautiful but it does not make my code simpler, so I'm happy to include it in the PR if you say that it is useful elsewhere or if you see that it permits to generalize other stuff in this PR that I am not aware of! In such case I'd be very interested to hear specifically what you have in mind!

Reply:
I am confident we'll want `derivation.comp` at some point (not necessarily here) so I've just created #7715 to propose it.

I haven't been able to convince myself what the best setup for your proposal is yet and it's possible this idea will not be of use. However I think it _should_ simplify your code since you would no longer have to prove `eval_map` and `eval_at`.

I'll have some more comments tomorrow.

Reply:
But the question I have in mind is: what is the correct abstraction for the various properties of `C^∞⟮I, M; 𝕜⟯`?

The theory you've developed requires the following properties:

- `C^∞⟮I, M; 𝕜⟯` is a `𝕜`-algebra,
- for any `x : M` we have a morphism of `𝕜`-algebras: `C^∞⟮I, M; 𝕜⟯ →ₗ[𝕜] 𝕜`.

Everything follows from these. It's possible this is one abstraction too far but I'd like to think a little more.

Reply:
@ocfnash the reason why it does not make my code simpler is that in my situation type class inference cannot work for the module structure of `C^∞⟮I, M; 𝕜⟯` on `𝕜`, because it depends on the point `x : M`, so the module structure must be fed explicitly each time and hence I would still have to define `eval_at` this way:
```
/-- The evaluation is a linear map. -/
def eval_at (x : M) : (derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯) →ₗ[𝕜] point_derivation I M x :=
@linear_map.derivation.comp 𝕜 _ C^∞⟮I, M; 𝕜⟯ _ _ C^∞⟮I, M; 𝕜⟯ _ _ _ 𝕜 _
  (point_derivation.module I M x) _ _ (point_derivation.is_scalar_tower I M x)
  (smooth_function.eval_at I x)
 ```
but again, it totally makes sense to include this generalization if it can come it useful elsewhere

Reply:
Agreed. The typeclass problems you mention are the main concern I have. The obvious thing would be to use a type synonym but I haven't quite convinced myself that's the best way to go yet.